### PR TITLE
Reduce code and fix webhook cache latency

### PR DESCRIPTION
**Pull request by FireDefend**

### CODE REVIEW

**Comment by msftcoderdjw** [comment]
@FireDefend  I saw integration tests and suite tests failed in this PR. Do you want to debug the test failures before review?

File: k8s/apis/federation/v1/catalog_types.go [hunk 1]
```diff
@@ -35,6 +37,26 @@ type CatalogList struct {
 	Items           []Catalog `json:"items"`
 }
 
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
```
Comment (RemindD) [inline k8s/apis/federation/v1/catalog_types.go:41]: unnecessary subresource status and root declaration

File: k8s/apis/solution/v1/solution_webhook.go [hunk 2]
```diff
@@ -126,7 +128,7 @@ func (r *Solution) validateCreateSolution() error {
 
 func (r *Solution) validateUpdateSolution() error {
 	var solutions SolutionList
-	err := mySolutionClient.List(context.Background(), &solutions, client.InNamespace(r.Namespace), client.MatchingFields{".spec.displayName": r.Spec.DisplayName})
+	err := mySolutionClient.List(context.Background(), &solutions, client.InNamespace(r.Namespace), client.MatchingFields{".spec.displayName": r.Spec.DisplayName}, client.Limit(2))
```
Comment (RemindD) [inline k8s/apis/solution/v1/solution_webhook.go:131]: why limit is 2 here?

File: k8s/apis/workflow/v1/campaign_webhook.go [hunk 3]
```diff
@@ -161,3 +162,17 @@ func (r *Campaign) validateRootResource() *field.Error {
 
 	return nil
 }
+
+func (r *CampaignContainer) ValidateDelete() (admission.Warnings, error) {
+	campaignlog.Info("validate delete solution container", "name", r.Name)
```
Comment (RemindD) [inline k8s/apis/workflow/v1/campaign_webhook.go:167]: typo

File: k8s/apis/solution/v1/solution_webhook.go (hunk 2 above)
Comment (FireDefend) [reply k8s/apis/solution/v1/solution_webhook.go:131]: this is an update check, the result may contain itself

File: k8s/apis/federation/v1/catalog_webhook.go [hunk 4]
```diff
@@ -247,3 +248,17 @@ func (r *Catalog) validateRootResource() *field.Error {
 
 	return nil
 }
+
+func (r *CatalogContainer) ValidateDelete() (admission.Warnings, error) {
+	cataloglog.Info("validate delete solution container", "name", r.Name)
```
[the remark this answers is not in the document]
Comment (FireDefend) [reply k8s/apis/federation/v1/catalog_webhook.go]: fixed, thanks for reminding

File: k8s/apis/federation/v1/catalog_types.go (hunk 1 above)
Comment (FireDefend) [reply k8s/apis/federation/v1/catalog_types.go]: fixed, thanks for reminding

File: k8s/apis/workflow/v1/campaign_webhook.go (hunk 3 above)
Comment (FireDefend) [reply k8s/apis/workflow/v1/campaign_webhook.go]: fixed, thanks for reminding

**Comment by msftcoderdjw** [comment]
@FireDefend please take a look at the test failures.

File: k8s/apis/federation/v1/catalog_types.go (hunk 1 above)
Comment (RemindD) [reply k8s/apis/federation/v1/catalog_types.go]: Looks like root=true is needed. Sorry for confusion.

File: k8s/apis/federation/v1/catalog_webhook.go [hunk 5]
```diff
@@ -68,20 +65,25 @@ func (r *Catalog) Default() {
...
 		var catalogContainer CatalogContainer
-		err := myCatalogClient.Get(context.Background(), client.ObjectKey{Name: r.Spec.RootResource, Namespace: r.Namespace}, &catalogContainer)
+		err := myCatalogReaderClient.Get(context.Background(), client.ObjectKey{Name: r.Spec.RootResource, Namespace: r.Namespace}, &catalogContainer)
 		if err != nil {
 			cataloglog.Error(err, "failed to get catalog container", "name", r.Spec.RootResource)
 		} else {
 			ownerReference := metav1.OwnerReference{
-				APIVersion: catalogContainer.APIVersion,
-				Kind:       catalogContainer.Kind,
+				APIVersion: GroupVersion.String(), //catalogContainer.APIVersion
```
Comment (msftcoderdjw) [inline k8s/apis/federation/v1/catalog_webhook.go:73]: Still use catalogContainer.APIVersion and catalogContainer.Kind.

File: k8s/apis/solution/v1/solution_webhook.go [hunk 6]
```diff
@@ -64,15 +62,20 @@ func (r *Solution) Default() {
 			solutionlog.Error(err, "failed to get solution container", "name", r.Spec.RootResource)
 		} else {
 			ownerReference := metav1.OwnerReference{
-				APIVersion: solutionContainer.APIVersion,
-				Kind:       solutionContainer.Kind,
+				APIVersion: GroupVersion.String(),
```
Comment (msftcoderdjw) [inline k8s/apis/solution/v1/solution_webhook.go:65]: Seems these two lines are not needed.

**Review by msftcoderdjw — APPROVED** [review]

**Pull request merged**